flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java

```diff
@@ -139,13 +144,43 @@ public boolean acknowledgeTask(
 			
 			ExecutionVertex vertex = notYetAcknowledgedTasks.remove(attemptID);
 			if (vertex != null) {
-				if (state != null) {
-					collectedStates.add(new StateForTask(
-							state,
-							stateSize,
-							vertex.getJobvertexId(),
+				if (state != null || kvState != null) {
+
+					JobVertexID jobVertexID = vertex.getJobvertexId();
+
+					StateForTaskGroup taskGroupState;
+
+					if (taskGroupStates.containsKey(jobVertexID)) {
+						taskGroupState = taskGroupStates.get(jobVertexID);
+					} else {
+						taskGroupState = new StateForTaskGroup(jobVertexID, vertex.getTotalNumberOfParallelSubtasks());
+						taskGroupStates.put(jobVertexID, taskGroupState);
+					}
+
+					long timestamp = System.currentTimeMillis() - checkpointTimestamp;
+
+					if (state != null) {
+						taskGroupState.putState(
 							vertex.getParallelSubtaskIndex(),
-							System.currentTimeMillis() - checkpointTimestamp));
+							new StateForTask(
+								state,
+								stateSize,
+								timestamp
+							)
+						);
+					}
+
+					if (kvState != null) {
+						for (Map.Entry<Integer, SerializedValue<StateHandle<?>>> entry : kvState.entrySet()) {
+							taskGroupState.putKvState(
+								entry.getKey(),
+								new KvStateForTasks(
+									entry.getValue(),
+									0L,
```

Comment:
Because at the moment the key-value state is still contained in the `state` object. This will be changed with the next PR introducing key groups for key-value state.